cpp/tests/utilities/mg_test_utilities.cu

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright (c) 2021-, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <experimental/graph.hpp>
+#include <experimental/detail/graph_utils.cuh>
+#include <experimental/graph_functions.hpp>
+#include <utilities/shuffle_comm.cuh>
+
+#include <utilities/mg_test_utilities.hpp>
+
+
+namespace cugraph {
+namespace test {
+
+// MG test utility which returns a MG graph_t instance constructed from a
+// edgelist (currently assumed to be generated from reading a .mtx file).  The
+// data is partitioned, shuffled, and renumbered based on the current GPU rank
+// before being passed to the graph_t ctor.
+//
+template <typename vertex_t, typename edge_t, typename weight_t, bool store_transposed>
+cugraph::experimental::graph_t<vertex_t, edge_t, weight_t, store_transposed, true> // multi_gpu=true
+create_graph_for_gpu(raft::handle_t& handle,
+                     edgelist_from_market_matrix_file_t<vertex_t, weight_t> edgelist_from_mm,
+                     bool input_is_weighted) {
+
+   const auto &comm = handle.get_comms();
+   auto& row_comm = handle.get_subcomm(cugraph::partition_2d::key_naming_t().row_name());
+   auto& col_comm = handle.get_subcomm(cugraph::partition_2d::key_naming_t().col_name());
+
+   int my_rank = comm.get_rank();
+
+   // Create a edge_gpu_identifier, which will be used by the individual jobs
+   // to identify if a edge belongs to a particular GPU/job.
+   //cugraph::experimental::detail::compute_gpu_id_from_edge_t<vertex_t> edge_gpu_identifier{false, comm.get_size(), row_comm.get_size(), col_comm.get_size()};
+   cugraph::experimental::detail::compute_gpu_id_from_vertex_t<vertex_t> vertex_gpu_identifier{comm.get_size()};
```

Comment:
compute_gpu_id_from_edge_t<vertex_t> should be used here.

Reply:
See https://github.com/rapidsai/cugraph/blob/branch-0.18/cpp/src/experimental/coarsen_graph.cu#L322 for an example.

Reply:
And you don't really need to call sort_and_shuffle_values here. This is for the case when the edges are partitioned in multiple GPUs. Here, every GPU holds the entire set of edges.

Also, note that edgelist.h_cols (and h_rows) are std::vector objects not rmm::device_uvector objects. h_cols.data() returns a pointer to host memory, the shuffle function expects device memory iterators.

In this case, you scan the entire edge list, apply the key_func (see the link above), and keep the edges with key_func() == comm.get_rank() (you may use remove_if).

Reply:
I think this is resolved in the latest commit.